src/main/fc/rc.c

```diff
@@ -664,49 +664,88 @@ FAST_CODE_NOINLINE void calculateFeedforward(const pidRuntime_t *pid, flight_dyn
+}
+
+FAST_CODE void writeSetpointAxis(int axis, float angleRate) {
+    rawSetpoint[axis] = constrainf(angleRate, -1.0f * currentControlRateProfile->rate_limit[axis], 1.0f * currentControlRateProfile->rate_limit[axis]);
+    DEBUG_SET(DEBUG_ANGLERATE, axis, angleRate);
+}
+
 FAST_CODE void processRcCommand(void)
 {
     if (isRxDataNew) {
         maxRcDeflectionAbs = 0.0f;
-        for (int axis = FD_ROLL; axis <= FD_YAW; axis++) {
 
-            float angleRate;
-            
+        float angleRateRoll = 0;
+        float angleRatePitch = 0;
+        float angleRateYaw = 0;
+
+        // Apply rates and expo for yaw axis independently
 #ifdef USE_GPS_RESCUE
-            if ((axis == FD_YAW) && FLIGHT_MODE(GPS_RESCUE_MODE)) {
-                // If GPS Rescue is active then override the setpointRate used in the
-                // pid controller with the value calculated from the desired heading logic.
-                angleRate = gpsRescueGetYawRate();
-                // Treat the stick input as centered to avoid any stick deflection base modifications (like acceleration limit)
-                rcDeflection[axis] = 0;
-                rcDeflectionAbs[axis] = 0;
-            } else
+        if (FLIGHT_MODE(GPS_RESCUE_MODE)) {
+            // If GPS Rescue is active then override the setpointRate used in the
+            // pid controller with the value calculated from the desired heading logic.
+            angleRateYaw = gpsRescueGetYawRate();
+            // Treat the stick input as centered to avoid any stick deflection base modifications (like acceleration limit)
+            rcDeflection[FD_YAW] = 0;
+            rcDeflectionAbs[FD_YAW] = 0;
+        } else
 #endif
-            {
-                // scale rcCommandf to range [-1.0, 1.0]
-                float rcCommandf;
-                if (axis == FD_YAW) {
-                    rcCommandf = rcCommand[axis] / rcCommandYawDivider;
-                } else {
-                    rcCommandf = rcCommand[axis] / rcCommandDivider;
-                }
-
-                rcDeflection[axis] = rcCommandf;
-                const float rcCommandfAbs = fabsf(rcCommandf);
-                rcDeflectionAbs[axis] = rcCommandfAbs;
-                maxRcDeflectionAbs = fmaxf(maxRcDeflectionAbs, rcCommandfAbs);
+        {
+            preProcessRcCommandAxis(FD_YAW); // Yaw normalize rcCommand [-1.0, 1.0], store in rcDeflection, rcDeflectionAbs, maxRcDeflectionAbs
+            angleRateYaw = applyRates(FD_YAW, rcDeflection[FD_YAW], rcDeflectionAbs[FD_YAW]);
+        }
 
-                angleRate = applyRates(axis, rcCommandf, rcCommandfAbs);
-            }
+        // Apply rates for pitch and roll
+        preProcessRcCommandAxis(FD_ROLL); // normalize rcCommand [-1.0, 1.0], store in rcDeflection, rcDeflectionAbs, maxRcDeflectionAbs
+        preProcessRcCommandAxis(FD_PITCH); // normalize rcCommand [-1.0, 1.0], store in rcDeflection, rcDeflectionAbs, maxRcDeflectionAbs
+#ifdef USE_VECTOR_EXPO
```

Comment:
this math probably could be done in a separate function for the same of cleaner code?

Reply:
Sure, for now I'd say the most important thing is testing it and seeing how it feels and if it would be worth implementing.